### PR TITLE
Handle Empty Sequence In Config

### PR DESCRIPTION
**Pull request by adamgfraser**
If we have a `ConfigProvider` like `Map("key" -> "")` and the `Config` is `Config.chunkOf(Config.string).nested("key")` currently we return a chunk with a single element containing the empty string. This PR changes that behavior to instead return an empty chunk.

One disadvantage of this change is that it makes loading a configuration with a sequence containing a single empty string unrepresentable, but that seems much better than the current behavior where an empty sequence is unrepresentable. Unfortunately it does not seem possible to support both without introducing an additional operator which seems potentially excessive for a somewhat degenerate case.

### CODE REVIEW

**Comment by afsalthaj** [comment]
There are a few subtle cases that's missing in this PR.

Example we need this to be working

```
      test("empty list with product") {
        val configProvider =
          ConfigProvider.fromMap(
            Map(
              "departments" -> ""
            )
          )

        val member = Config.int("age").zip(Config.string("name"))

        val config = Config.listOf("departments", member)

        for {
          result <- configProvider.load(config)
        } yield assertTrue(result == Nil)
      } 

```

I didn't know you were already in progress, but I have a draft PR (now can be used as a reference) that handles the above scenario. The draft PR has a failing test though (which is `empty list within indexed list`)

https://github.com/zio/zio/pull/7842

**Comment by adamgfraser** [comment]
I think we are seeing some of the disadvantages of attempting to support both "traditional" and "indexed" structures here.

Previously, we used the presence of an index following a key to indicate that this was an indexed configuration provider and we should process it accordingly.

However, now we have an ambiguity because if we have a case like `employees` we don't know whether that is indicating (1) we are in an indexed configuration provider and this is just trying to tell us there are no indices or (2) we are in a traditional configuration provider and need to process it accordingly.

Will think about this.

**Comment by afsalthaj** [comment]
The indexed sequence is superset of any situation in my opinion.



```scala
 "ages" -> "1, 2, 3"
```

can be 

```scala
"ages[0]" -> "1"
"ages[1]" -> "2"

```

Once transformed everything works as before (i.e, enumerate children and nesting etc, i.e, almost zero code change) 

Another complex example:


```scala
  Map( 
      "employee[0].age" -> 1
      "employee[1].age" -> 2    
  )

```

is better than 

```scala
  Map( 
      "employee.ages" -> 1, 2
  )
```

The former can already represent the latter. 
That said, we shouldn't impose any rules how users encode this. For complex use case, I feel users going with indexed approach. Even if they don't prefer index, we can easily transform it as explained in the first example.


To sum up, this implies we transform values that are comma separate in values to indexed values (i.e, raw source to the initial structure in the above example), before feeding to the `loop` logic, where it only plays with index. This works with any complex sources, consistent, and avoid splitting of values when doing `parsePrimitive` and makes the code cleaner as far as I see.

**Comment by adamgfraser** [comment]
I think you are right that the indexed format is superior to the traditional format in that it can represent more complex structures of data than can be represented in the traditional format.

However, I think the traditional format is more user friendly and fits user expectations when configuration information is defined directly, for example in system properties or environment variables. For instance if I wanted to define some ages in system properties I would do it like this:

```
`age -> 1, 2, 3`
```

Not:

```
age[0] -> 1
age[1] -> 2
age[2] -> 3
```

Therefore, I think the indexed format will likely be more of an internal format for representing data loaded from more complex configuration structures such as JSON, which is an extremely important use case.

In addition, unfortunately I do not think it is possible to mechanically convert from a traditional format to the indexed one. Taking your example, if we have:

```scala
  Map( 
      "employee.age" -> 1, 2
  )
```

The straightforward translation to an indexed format would be:

```scala
  Map( 
      "employee.age[0]" -> 1
      "employee.age[1]" -> 2
  )
```

But this is wrong! We are supported to have a list of employees, not a list of ages! And we don't know from the structure of the first data where the index is supposed to go. So these really are different formats that need to be handled differently.

**Comment by afsalthaj** [comment]
@adamgfraser 

Coming back to the original problem, how do we handle a empty list in a json, which calls for the usage of `IndexedFlat`

## Source:

```scala
  {
    departments: [] 
  }

```

or

```
  {
     departments : null
  }

```

What should be it's flattened form in order to work with `IndexedFlat` ?

May be the answer is this : https://github.com/zio/zio/pull/7843#discussion_r1111053238

**Comment by adamgfraser** [comment]
@afsalthaj Yes I think the first step is to separate the interpretation of "traditional" and "indexed" flat providers. Then the "indexed" format can represent an empty list however it wants.
